File: 2020/node/graph.ts
import { Queue } from "@datastructures-js/queue";

export function bfs<TNode, TKey>(
  start: TNode,
  adjacent: (n: TNode) => Iterable<TNode>,
  key: (n: TNode) => TKey,
  visit: (n: TNode) => void
) {
  const queue = new Queue<TNode>();
  const visited = new Set<TKey>();

  queue.enqueue(start);
  while (queue.size() > 0) {
    const node = queue.dequeue();
    if (visited.has(key(node))) {
      continue;
    }

    visited.add(key(node));
    visit(node);

    for (const adj of adjacent(node)) {
      if (!visited.has(key(adj))) {
        queue.enqueue(adj);
      }
    }
  }
}

File: 2021/utils/graph.ml
open! Base
open Core

let bfs m ~start ~adjacent =
  let visited = Hash_set.create m in
  let queue = Queue.create () in
  let () = Queue.enqueue queue start in
  let rec loop () =
    if Queue.is_empty queue then visited |> Hash_set.to_list
    else
      let p = Queue.dequeue_exn queue in
      if Hash_set.mem visited p then loop ()
      else
        let () = Hash_set.add visited p in
        let () =
          adjacent p visited
          |> List.filter ~f:(fun adj -> not (Hash_set.mem visited adj))
          |> List.iter ~f:(Queue.enqueue queue)
        in
        loop ()
  in
  loop ()
